Mustache: Render Map as JSON

tlrx opened the pull request:
~~This commit changes Mustache script engine so that every map is rendered as a JSON string (instead of default Map.toString())~~

This pull request adds two util functions to the Mustache templating engine:
- {{#toJson}}my_map{{/toJson}} to render a Map parameter as a JSON string
- {{#join}}my_iterable{{/join}} to render any iterable (including arrays) as a comma separated list of values like `1, 2, 3`. It's also possible de change the default delimiter (comma) to something else.

closes #18970

rjernst commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java:
```diff
@@ -40,6 +44,14 @@ public Object coerce(Object object) {
+                @SuppressWarnings("unchecked")
+                JsonMap map = new JsonMap((Map<String, Object>) object);
+                return map;
+            } catch (ClassCastException e) {
```
I don't think this can ever result in a `ClassCastException` because of type erasure.

rjernst commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java:
```diff
@@ -154,4 +166,59 @@ public boolean containsKey(Object key) {
+        public String toString() {
+            try (XContentBuilder builder = XContentBuilder.builder(XContentType.JSON.xContent())) {
+                return builder.map(internal).string();
+            } catch (IOException e) {
```
Why hide ioe, it should just propagate...

rjernst commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomReflectionObjectHandler.java:
```diff
@@ -154,4 +166,59 @@ public boolean containsKey(Object key) {
+                        // Ignore if it's not a number, we tried everything we could
+                    }
+                }
+            } else if (key instanceof Number) {
```
This is a String,Object map, how can key be a number?

tlrx commented:
I updated the code to make it work with the `{{#toJson}}ctx.payload.someMap{{/toJson}}` syntax. I also added a `{{#join}}ctx.payload.someIterable{{/join}}` to render things like comma separated values.

It would be great if you can review and test it.

@rjernst Thanks for your review! The pull request was marked as work in progress and I changed the whole implementation after discussing with @spinscale. I took your comment into consideration, I'd be glad if you find the time to review again.

rjernst commented:
I have no opinion on the technical side of this change. I'm happy this allows rendering json, but I also think the syntax looks funky based on other templating languages I have used in the past. I would have expected something simpler like `{{ctx.payload.someMap.asJson}}` or `{{asJson(ctx.payload.someMap)}}`

spinscale commented on modules/lang-mustache/src/test/java/org/elasticsearch/script/mustache/MustacheTests.java:
```diff
@@ -177,4 +184,182 @@ public void testSizeAccessForCollectionsAndArrays() throws Exception {
+
+        assertScript("{{#toJson}}.{{/toJson}}", ctx, equalTo("{\"ctx\":{\"name\":\"John Smith\",\"age\":42,\"height\":1.84}}"));
+        assertScript("{{#toJson}}ctx{{/toJson}}", ctx, equalTo("{\"name\":\"John Smith\",\"age\":42,\"height\":1.84}"));
+        assertScript("{{#toJson}}ctx.name{{/toJson}}", ctx, equalTo("\"John Smith\""));
```
does this make sense? is that valid JSON?

tlrx replied:
Not really. Let's change that so that it renders "John Smith".

spinscale commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/CustomMustacheFactory.java:
```diff
@@ -0,0 +1,223 @@
+        static final String CODE = "join";
+
+        public JoinerCode(TemplateContext tc, DefaultMustacheFactory df, Mustache mustache, String variable) {
+            super(tc, df, mustache, extractVariableName(CODE, mustache, tc));
```
`variable` unused?

tlrx replied:
I changed that.

spinscale commented:
I had a quick look but did not find any possibility for having arguments to configure the string joiner char instead of hardcoding it as a comma, that would be nice.

Apart from that there needs to be some documentation in the search template docs for those functions

tlrx commented:
Thanks @spinscale for the review. I updated the code, changing few things around and added the possibility to set a custom delimiter in the `join` function. Please let me know what you think.

> I have no opinion on the technical side of this change. I'm happy this allows rendering json, but I also think the syntax looks funky based on other templating languages I have used in the past. I would have expected something simpler like {{ctx.payload.someMap.asJson}} or {{asJson(ctx.payload.someMap)}}

@rjernst Yes, I don't have a strong feeling here. I agreed with you when I discussed with @spinscale about implementing a `toJSON()` (you called it `asJson`) method on arrays and maps. 

But the Mustache specs also propose wrapped functions section with `{{#wrapped}}...{{/wrapped}}` so I think having `{{#toJson}}...{{/toJson}}` or `{{#join}}...{{/join}}` syntax is OK. It also has the side effect to allow things like `{{#join}}{{#toJson}}terms{{/toJson}}{{/join}}` which would be more difficult to express using the other syntax.

spinscale commented on modules/lang-mustache/src/main/java/org/elasticsearch/script/mustache/MustacheScriptEngineService.java:
```diff
@@ -89,21 +89,19 @@ public MustacheScriptEngineService(Settings settings) {
+    }
+
+    private boolean isJsonEscapingEnabled(Map<String, String> params) {
+        switch (params.getOrDefault(CONTENT_TYPE_PARAM, JSON_CONTENT_TYPE)) {
```
`return JSON_CONTENT_TYPE.equals(params.getOrDefault(CONTENT_TYPE_PARAM, JSON_CONTENT_TYPE))`?

tlrx replied:
Sure, this switch statement is really ugly. I suppose my brain was in auto-pilot mode when writing this.

spinscale commented on docs/reference/search/search-template.asciidoc:
```diff
@@ -89,6 +89,89 @@ which is rendered as:
+    "query": {
+      "range": {
+        "born": {
+            "lte"   : "{{date.min}}",
```
I think `lte` and `gte` is switched here?

spinscale commented:
left two minor comments. I like it, very helpful for mustache users!

LGTM

tlrx commented:
Thanks @spinscale
